File: DataFeedViews/TransitionTypeGroupDataFeedView.vb
'************************************************************************************
' StochasticTime: A .NET class library for developing stochastic time models.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports SyncroSim.Core
Imports SyncroSim.Core.Forms
Imports System.Windows.Forms
Imports System.Reflection

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class TransitionTypeGroupDataFeedView

    Private m_Grid As BaseDataGridView

    Protected Overrides Sub InitializeView()

        MyBase.InitializeView()

        Dim v As DataFeedView = Me.Session.CreateMultiRowDataFeedView(Me.Project)
        Me.PanelMain.Controls.Add(v)

    End Sub

    Protected Overrides Sub Dispose(ByVal disposing As Boolean)

        If (disposing And Not Me.IsDisposed) Then

            If (Me.m_Grid IsNot Nothing) Then

                RemoveHandler Me.m_Grid.CellBeginEdit, AddressOf OnGridCellBeginEdit
                RemoveHandler Me.m_Grid.CellEndEdit, AddressOf OnGridCellEndEdit

            End If

            If components IsNot Nothing Then
                components.Dispose()
            End If

        End If

        MyBase.Dispose(disposing)

    End Sub

    Public Overrides Sub LoadDataFeed(dataFeed As Core.DataFeed)

        MyBase.LoadDataFeed(dataFeed)

        Dim v As MultiRowDataFeedView = CType(Me.PanelMain.Controls(0), MultiRowDataFeedView)
        Me.m_Grid = v.GridControl()
        v.LoadDataFeed(dataFeed, DATASHEET_TRANSITION_TYPE_GROUP_NAME)

        AddHandler Me.m_Grid.CellBeginEdit, AddressOf OnGridCellBeginEdit
        AddHandler Me.m_Grid.CellEndEdit, AddressOf OnGridCellEndEdit

    End Sub

    Public Overrides Sub EnableView(enable As Boolean)

        If (Me.PanelMain.Controls.Count > 0) Then

            Dim v As DataFeedView = CType(Me.PanelMain.Controls(0), DataFeedView)
            v.EnableView(enable)

        End If

    End Sub

    Private Sub OnGridCellBeginEdit(ByVal sender As Object, ByVal e As DataGridViewCellCancelEventArgs)

        If (e.ColumnIndex = Me.m_Grid.Columns(DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME).Index) Then

            Dim dgv As DataGridViewRow = Me.m_Grid.Rows(e.RowIndex)
            Dim ds As DataSheet = Me.Project.GetDataSheet(DATASHEET_TRANSITION_GROUP_NAME)
            Dim filter As String = "IsAuto IS NULL OR IsAuto=0"
            Dim dv As New DataView(ds.GetData(), filter, ds.DisplayMember, DataViewRowState.CurrentRows)
            Dim Cell As DataGridViewComboBoxCell = CType(dgv.Cells(DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME), DataGridViewComboBoxCell)

            Cell.DataSource = dv
            Cell.ValueMember = ds.ValueMember
            Cell.DisplayMember = ds.DisplayMember

        End If

    End Sub

    Private Sub OnGridCellEndEdit(ByVal sender As Object, ByVal e As DataGridViewCellEventArgs)

        If (e.ColumnIndex = Me.m_Grid.Columns(DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME).Index) Then

            Dim dgv As DataGridViewRow = Me.m_Grid.Rows(e.RowIndex)
            Dim Cell As DataGridViewComboBoxCell = CType(dgv.Cells(DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME), DataGridViewComboBoxCell)
            Dim Column As DataGridViewComboBoxColumn = CType(Me.m_Grid.Columns(DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME), DataGridViewComboBoxColumn)

            Cell.DataSource = Column.DataSource
            Cell.ValueMember = Column.ValueMember
            Cell.DisplayMember = Column.DisplayMember

        End If

    End Sub

End Class

File: DataFeedViews/TransitionTypeGroupDataFeedView.cs
﻿// ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Core;
using SyncroSim.Core.Forms;
using System.Windows.Forms;
using System.Reflection;
using System.Data;

namespace SyncroSim.STSim
{
    [ObfuscationAttribute(Exclude=true, ApplyToMembers=false)]
    internal partial class TransitionTypeGroupDataFeedView
    {
        public TransitionTypeGroupDataFeedView()
        {
            InitializeComponent();
        }

        private BaseDataGridView m_Grid;

        protected override void InitializeView()
        {
            base.InitializeView();

            DataFeedView v = this.Session.CreateMultiRowDataFeedView(this.Project);
            this.PanelMain.Controls.Add(v);
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing && !this.IsDisposed)
            {
                if (this.m_Grid != null)
                {
                    this.m_Grid.CellBeginEdit -= OnGridCellBeginEdit;
                    this.m_Grid.CellEndEdit -= OnGridCellEndEdit;
                }

                if (components != null)
                {
                    components.Dispose();
                }
            }

            base.Dispose(disposing);
        }

        public override void LoadDataFeed(Core.DataFeed dataFeed)
        {
            base.LoadDataFeed(dataFeed);

            MultiRowDataFeedView v = (MultiRowDataFeedView)this.PanelMain.Controls[0];
            this.m_Grid = v.GridControl;
            v.LoadDataFeed(dataFeed, Strings.DATASHEET_TRANSITION_TYPE_GROUP_NAME);

            this.m_Grid.CellBeginEdit += OnGridCellBeginEdit;
            this.m_Grid.CellEndEdit += OnGridCellEndEdit;
        }

        public override void EnableView(bool enable)
        {
            if (this.PanelMain.Controls.Count > 0)
            {
                DataFeedView v = (DataFeedView)this.PanelMain.Controls[0];
                v.EnableView(enable);
            }
        }

        private void OnGridCellBeginEdit(object sender, DataGridViewCellCancelEventArgs e)
        {
            if (e.ColumnIndex == this.m_Grid.Columns[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME].Index)
            {
                DataGridViewRow dgv = this.m_Grid.Rows[e.RowIndex];
                DataSheet ds = this.Project.GetDataSheet(Strings.DATASHEET_TRANSITION_GROUP_NAME);
                string filter = "IsAuto IS NULL OR IsAuto=0";
                DataView dv = new DataView(ds.GetData(), filter, ds.DisplayMember, DataViewRowState.CurrentRows);
                DataGridViewComboBoxCell Cell = (DataGridViewComboBoxCell)dgv.Cells[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME];

                Cell.DataSource = dv;
                Cell.ValueMember = ds.ValueMember;
                Cell.DisplayMember = ds.DisplayMember;
            }
        }

        private void OnGridCellEndEdit(object sender, DataGridViewCellEventArgs e)
        {
            if (e.ColumnIndex == this.m_Grid.Columns[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME].Index)
            {
                DataGridViewRow dgv = this.m_Grid.Rows[e.RowIndex];
                DataGridViewComboBoxCell Cell = (DataGridViewComboBoxCell)dgv.Cells[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME];
                DataGridViewComboBoxColumn Column = (DataGridViewComboBoxColumn)this.m_Grid.Columns[Strings.DATASHEET_TRANSITION_GROUP_ID_COLUMN_NAME];

                Cell.DataSource = Column.DataSource;
                Cell.ValueMember = Column.ValueMember;
                Cell.DisplayMember = Column.DisplayMember;
            }
        }
    }
}